src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ServiceChannelFactory.cs

```diff
@@ -390,6 +395,33 @@ internal static ServiceChannel GetServiceChannel(object transparentProxy)
                 return null;
         }
 
+        private async Task OnCloseAsyncInternal(TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            while (true)
+            {
+                int count;
+                IChannel channel;
+                lock (ThisLock)
+                {
+                    count = _channelsList.Count;
+                    if (count == 0)
+                        return;
+                    channel = _channelsList[0];
+                }
+
+                IAsyncCommunicationObject asyncChannel = channel as IAsyncCommunicationObject;
+                if (asyncChannel != null)
+                {
+                    await asyncChannel.CloseAsync(timeoutHelper.RemainingTime());
+                }
+                else
+                {
+                    await Task.Factory.FromAsync(channel.BeginClose, channel.EndClose, TaskCreationOptions.None);
```

Comment:
Actually, I believe you can pass a parameter to the FromAsync method which gets passed to the first delegate. It's a generic variant of FromAsync where the generic type parameter is the type of the passed argument.

---
In reply to: [60984917](https://github.com/dotnet/wcf/pull/1071#discussion_r60984917) [](ancestors = 60984917,60769739)